docs/configuration.md
# Configuration

`config/initializers/react_on_rails_pro.rb`  

1. You don't need to create a initializer if you are satisfied with the defaults as described below.
1. Values beginning with `renderer` pertain only to using an external rendering server. You will need to ensure these values are consistent with your configuration for the external rendering server, as given in [JS configuration](https://www.shakacode.com/react-on-rails-pro/docs/node-renderer/js-configuration/)
1. `config.prerender_caching` works for standard mini_racer server rendering and using an external rendering server.

## Example of Configuration

Also see [spec/dummy/config/initializers/react_on_rails_pro.rb](https://github.com/shakacode/react_on_rails_pro/blob/master/spec/dummy/config/initializers/react_on_rails_pro.rb) for how the testing app is setup.

The below example is a typical production setup, using the separate `NodeRenderer`, where development takes the defaults when the ENV values are not specified.

```ruby
ReactOnRailsPro.configure do |config|
  # If true, then capture timing of React on Rails Pro calls including server rendering and
  # component rendering.
  # Default for `tracing` is false.
  config.tracing = true

  # Array of globs to find any files for which changes should bust the fragment cache for
  # cached_react_component and cached_react_component_hash. This should include any files used to
  # generate the JSON props, webpack and/or webpacker configuration files, and npm package lockfiles.
  # Default for `dependency_globs` is an empty array
  config.dependency_globs = [ File.join(Rails.root, "app", "views", "**", "*.jbuilder") ]

  # Array of globs to exclude from config.dependency_globs for ReactOnRailsPro cache key hashing
  # Default for `excluded_dependency_globs` is an empty array
  config.excluded_dependency_globs = [ File.join(Rails.root, "app", "views", "**", "dont_hash_this.jbuilder") ]

  # Remote bundle caching saves deployment time by caching bundles.
  # See /docs/bundle-caching.md for usage and an example of a module called S3BundleCacheAdapter.
  config.remote_bundle_cache_adapter = nil
  
  # ALL OPTIONS BELOW ONLY APPLY IF SERVER RENDERING

  # If true, then cache the evaluation of JS for prerendering using the standard Rails cache.
  # Applies to all rendering engines.
  # Default for `prerender_caching` is false.  
  config.prerender_caching = true

  # Retry request in case of time out on the node-renderer side
  # 5 - default, if not specified
  # 0 - no retry
  config.renderer_request_retry_limit = 5

  # NodeRenderer is for a renderer that is stateless. It does not need restarting when the JS bundles
  # are updated. It is the only custom renderer currently supported. Leave blank to use the standard
  # mini_racer rendering. Other option is NodeRenderer
  # Default for `server_renderer` is "ExecJS"
  config.server_renderer = "NodeRenderer"

  # React on Rails Node Renderer now support render functions returning promises! To enable this optional functionality,
  # toggle the following option.
  # Default is false.
  config.rendering_returns_promises = false

  # If you're using the NodeRenderer, a value of true allows errors to be thrown from the bundle
  # code for SSR so that an error tracking system on the NodeRender can use the exceptions.
  # If you are using ExecJS as your rendering method, set this to false.
  # Default is true.
  config.throw_js_errors = true

  # You may provide a password and/or a port that will be sent to renderer for simple authentication.
  # `https://:<password>@url:<port>`. For example: https://:myPassword1@renderer:3800. Don't forget
  # the leading `:` before the password. Your password must also not contain certain characters that
  # would break calling URI(config.renderer_url). This includes: `@`, `#`, '/'.
  # **Note:** Don't forget to set up **SSL** connection (https) otherwise password will useless
  # since it will be easy to intercept it.
  # If you provide an ENV value (maybe only for production) and there is no value, then you get the default.
  # Default for `renderer_url` is "http://localhost:3800".
  config.renderer_url = ENV["RENDERER_URL"]

  # If you don't want to worry about special characters in your password within the url, use this config value
  # Default for `renderer_password` is ""
  # config.renderer_password = ENV["RENDERER_PASSWORD"]

  # Set the `ssr_timeout` configuration so the Rails server will not wait more than this many seconds
  # for a SSR request to return once issued. 
  config.ssr_timeout = 5
  
  # If false, then crash if no backup rendering when the remote renderer is not available
  # Can be useful to set to false in development or testing to make sure that the remote renderer
  # works and any non-availability of the remote renderer does not just do ExecJS.
  # Suggest setting this to false if the SSR JS code cannot run in ExecJS
  # Default for `renderer_use_fallback_exec_js` is false.
  config.renderer_use_fallback_exec_js = false

  # The maximum size of the http connection pool,
  # Set +pool_size+ to limit the maximum number of connections allowed.
  # Defaults to 1/4 the number of allowed file handles.  You can have no more
  # than this many threads with active HTTP transactions.
  # Default for `renderer_http_pool_size` is 10
  config.renderer_http_pool_size = 10

  # Seconds to wait for an available connection before a timeout error is raised
  # Default for `renderer_http_pool_timeout` is 5
  config.renderer_http_pool_timeout = 5

  # warn_timeout  - Displays an error message if a request takes longer than the given time in seconds
  # (used to give hints to increase the pool size). Default is 0.25
  config.renderer_http_pool_warn_timeout = 0.25 # seconds

  # Snippet of JavaScript to be run right at the beginning of the server rendering process. The code
  # to be executed must either be self contained or reference some globally exposed module.  
  # For example, suppose that we had to call `SomeLibrary.clearCache()`between every call to server
  # renderer to ensure no leakage of state between calls. Note, SomeLibrary needs to be globally
  # exposed in the server rendering webpack bundle. This code is visible in the tracing of the calls
  # to do server rendering. Default is nil.
  config.ssr_pre_hook_js = "SomeLibrary.clearCache();"

  # When using the Node Renderer, you may require some extra assets in addition to the bundle.
  # The assets_to_copy option allows the Node Renderer to have assets copied at the end of
  # the assets:precompile task or directly by the
  # react_on_rails_pro:copy_assets_to_vm_renderer tasks.
  # These assets are also transferred any time a new bundle is sent from Rails to the renderer.
  # The value should be a file_path or an Array of file_paths. The files should have extensions
  # to resolve the content types, such as "application/json".
  config.assets_to_copy = [
     Rails.root.join("public", "webpack", Rails.env, "loadable-stats.json"),
     Rails.root.join("public", "webpack", Rails.env, "manifest.json")
  ]
end
```


NEWS.md
# NEWS

*A history of the news. A few bullets at the top will also show on the [README.md](./README.md).*

* **October 14, 2020**: [RUBY ROGUES
RR 474: React on Rails V12 – Don’t Shave That Yak! with Justin Gordon](https://devchat.tv/ruby-rogues/rr-474-react-on-rails-v12-dont-shave-that-yak-with-justin-gordon/).

* **October 1, 2020**: See the [spec/dummy](https://github.com/shakacode/react_on_rails/tree/master/spec/dummy) example repo for a simple configuration of webpack via the rails/webpacker gem
that supports SSR.
* **August 2, 2020**: See the example repo of [React on Rails Tutorial With SSR, HMR fast refresh, and TypeScript](https://github.com/shakacode/react_on_rails_demo_ssr_hmr) for a new way to setup the creation of your SSR bundle with `rails/webpacker`.
* July 8, 2020: Release **v12**.
  1. **React Hooks Support** for top level components
  2. **Typescript bindings**
  3. **rails/webpacker** "just works" with React on Rails by default.
  4. i18n support for generating a JSON file rather than a JS file.
* 2018-02-27: **Version 10.1.2** Supports the React API for ReactDOM.hydrate.
* 2017-09-06: **VERSION 9.0.0 shipped!** This version depends on Webpacker directly. See [Upgrading React on Rails](./docs/guides/upgrading-react-on-rails.md) for more concise instructions on upgrading.
* Always see the [CHANGELOG.md](./CHANGELOG.md) for the latest project changes.
* [VERSION 8.1.0](https://rubygems.org/gems/react_on_rails/) shipped with [webpacker_lite](https://github.com/shakacode/webpacker_lite) (soon [**webpacker**](https://github.com/rails/webpacker/issues/464#issuecomment-310986140) support! [react-webpack-rails-tutorial PR #395](https://github.com/shakacode/react-webpack-rails-tutorial/pull/395) shows the changes needed to migrate from the Asset Pipeline to Webpacker Lite. For more information, see my article: [Webpacker Lite: Why Fork Webpacker?](https://blog.shakacode.com/webpacker-lite-why-fork-webpacker-f0a7707fac92). Per recent discussions, we [will merge Webpacker Lite changes back into Webpacker](https://github.com/rails/webpacker/issues/464#issuecomment-310986140). There's no reason to wait for this. The upgrade will eventually be trivial.
* 2017-04-25: 7.0.0 Shipped! Performance improvements! Please upgrade! Only "breaking" change is that you have to update both the node module and the Ruby gem. 
* 2017-04-09: 8.0.0 beta work to include webpacker_lite gem has begun. See [#786](https://github.com/shakacode/react_on_rails/issues/786).
* 2017-04-03: 6.9.3 Released! Props rendered in JSON script tag. Page size is smaller now due to less escaping!
* 2017-03-06: Updated to Webpack v2!
* 2017-03-02: Demo of internationalization (i18n) is live at [reactrails.com](https://www.reactrails.com/). Docs [here](docs/guides/i18n.md).
* 2017-02-28: See [discussions here on Webpacker](https://github.com/rails/webpacker/issues/139) regarding how Webpacker will allow React on Rails to avoid using the asset pipeline in the near future.
* 2017-02-28: Upgrade to Webpack v2 or use the `--bail` option in your webpack script for test and production builds. See the discussion on [PR #730](https://github.com/shakacode/react_on_rails/pull/730).
* 2016-11-03: Spoke at [LA Ruby: "React on Rails: Why, What, and How?"](http://www.meetup.com/laruby/events/234825187/). [Video and pictures in this article](https://blog.shakacode.com/my-react-on-rails-talk-at-the-la-ruby-rails-meetup-november-10-2016-eaaa83aff800#.ej6h4eglp).
* 2016-12-20: New Video on Egghead.io: [Creating a component with React on Rails](https://egghead.io/lessons/react-creating-a-component-with-react-on-rails)
* 2016-11-03: Spoke at [LA Ruby, 7pm, Thursday, November 10 in Venice, CA: "React on Rails: Why, What, and How?"](http://www.meetup.com/laruby/events/234825187/). [Video and pictures in this article](https://blog.shakacode.com/my-react-on-rails-talk-at-the-la-ruby-rails-meetup-november-10-2016-eaaa83aff800#.ej6h4eglp).
* 2016-08-27: We now have a [Documentation Gitbook](https://shakacode.gitbooks.io/react-on-rails/content/) for improved readability & reference.
* 2016-08-21: v6.1 ships with several new features and bug fixes. See the [Changelog](CHANGELOG.md).
* 2016-07-28: If you're doing server rendering, be sure to use mini\_racer! See [issues/428](https://github.com/shakacode/react_on_rails/issues/428). It's supposedly much faster than `therubyracer`.

* 2016-08-27: We now have a [Documentation Gitbook](https://shakacode.gitbooks.io/react-on-rails/content/) for improved readability & reference.
* 2016-08-21: v6.1 ships with several new features and bug fixes. See the [Changelog](CHANGELOG.md).
* 2016-06-13: 6.0.4 shipped with a critical fix regarding a missing polyfill for `clearTimeout`, used by babel-polyfill.
* 2016-06-06: 6.0.2 shipped with a critical fix if you are fragment caching the server generated React.
* 2016-05-24: 6.0.0 Released! Simplified generator and install process! See the [CHANGELOG.md](./CHANGELOG.md) for details.
* 2016-04-08: 5.2.0 Released! Support for React 15.0 and updates to the Generator.
* 2016-03-18: [Slides on React on Rails](http://www.slideshare.net/justingordon/react-on-rails-v4032).
* 2016-03-17: **4.0.3** Shipped! Includes using the new Heroku buildpack steps, several smaller changes detailed in the [CHANGELOG.md](./CHANGELOG.md). 
* 2016-03-17: **4.0.2** Shipped! Includes using the new Heroku buildpack steps.
  * Better support for hot reloading of assets from Rails with new helpers and updates to the sample testing app, [spec/dummy](spec/dummy).
  * Better support for Turbolinks 5.
  * Controller rendering of shared redux stores and ability to render store data at bottom of HTML page.
  * See [#311](https://github.com/shakacode/react_on_rails/pull/311/files).
  * Some breaking changes! See [CHANGELOG.md](./CHANGELOG.md) for details.

react_on_rails_pro/lib/react_on_rails_pro/configuration.rb
# frozen_string_literal: true

module ReactOnRailsPro
  def self.configure
    yield(configuration)
    configuration.setup_config_values
  end

  def self.configuration
    @configuration ||= Configuration.new(
      prerender_caching: Configuration::DEFAULT_PRERENDER_CACHING,
      server_renderer: Configuration::DEFAULT_RENDERER_METHOD,
      renderer_url: Configuration::DEFAULT_RENDERER_URL,
      renderer_use_fallback_exec_js: Configuration::DEFAULT_RENDERER_FALLBACK_EXEC_JS,
      renderer_http_pool_size: Configuration::DEFAULT_RENDERER_HTTP_POOL_SIZE,
      renderer_http_pool_timeout: Configuration::DEFAULT_RENDERER_HTTP_POOL_TIMEOUT,
      renderer_http_pool_warn_timeout: Configuration::DEFAULT_RENDERER_HTTP_POOL_WARN_TIMEOUT,
      renderer_password: nil,
      tracing: Configuration::DEFAULT_TRACING,
      dependency_globs: Configuration::DEFAULT_DEPENDENCY_GLOBS,
      excluded_dependency_globs: Configuration::DEFAULT_EXCLUDED_DEPENDENCY_GLOBS,
      remote_bundle_cache_adapter: Configuration::DEFAULT_REMOTE_BUNDLE_CACHE_ADAPTER,
      ssr_timeout: Configuration::DEFAULT_SSR_TIMEOUT,
      ssr_pre_hook_js: nil,
      assets_to_copy: nil,
      renderer_request_retry_limit: Configuration::DEFAULT_RENDERER_REQUEST_RETRY_LIMIT,
      throw_js_errors: Configuration::DEFAULT_THROW_JS_ERRORS,
      rendering_returns_promises: Configuration::DEFAULT_RENDERING_RETURNS_PROMISES,
      profile_server_rendering_js_code: Configuration::DEFAULT_PROFILE_SERVER_RENDERING_JS_CODE,
      raise_non_shell_server_rendering_errors: Configuration::DEFAULT_RAISE_NON_SHELL_SERVER_RENDERING_ERRORS
    )
  end

  class Configuration
    DEFAULT_RENDERER_URL = "http://localhost:3800"
    DEFAULT_RENDERER_METHOD = "ExecJS"
    DEFAULT_RENDERER_FALLBACK_EXEC_JS = true
    DEFAULT_RENDERER_HTTP_POOL_SIZE = 10
    DEFAULT_RENDERER_HTTP_POOL_TIMEOUT = 5
    DEFAULT_RENDERER_HTTP_POOL_WARN_TIMEOUT = 0.25
    DEFAULT_SSR_TIMEOUT = 5
    DEFAULT_PRERENDER_CACHING = false
    DEFAULT_TRACING = false
    DEFAULT_DEPENDENCY_GLOBS = [].freeze
    DEFAULT_EXCLUDED_DEPENDENCY_GLOBS = [].freeze
    DEFAULT_REMOTE_BUNDLE_CACHE_ADAPTER = nil
    DEFAULT_RENDERER_REQUEST_RETRY_LIMIT = 5
    DEFAULT_THROW_JS_ERRORS = true
    DEFAULT_RENDERING_RETURNS_PROMISES = false
    DEFAULT_PROFILE_SERVER_RENDERING_JS_CODE = false
    DEFAULT_RAISE_NON_SHELL_SERVER_RENDERING_ERRORS = false

    attr_accessor :renderer_url, :renderer_password, :tracing,
                  :server_renderer, :renderer_use_fallback_exec_js, :prerender_caching,
                  :renderer_http_pool_size, :renderer_http_pool_timeout, :renderer_http_pool_warn_timeout,
                  :dependency_globs, :excluded_dependency_globs, :rendering_returns_promises,
                  :remote_bundle_cache_adapter, :ssr_pre_hook_js, :assets_to_copy,
                  :renderer_request_retry_limit, :throw_js_errors, :ssr_timeout,
                  :profile_server_rendering_js_code, :raise_non_shell_server_rendering_errors

    # rubocop:disable Metrics/AbcSize
    def initialize(renderer_url: nil, renderer_password: nil, server_renderer: nil,
                   renderer_use_fallback_exec_js: nil, prerender_caching: nil,
                   renderer_http_pool_size: nil, renderer_http_pool_timeout: nil,
                   renderer_http_pool_warn_timeout: nil, tracing: nil,
                   dependency_globs: nil, excluded_dependency_globs: nil, rendering_returns_promises: nil,
                   remote_bundle_cache_adapter: nil, ssr_pre_hook_js: nil, assets_to_copy: nil,
                   renderer_request_retry_limit: nil, throw_js_errors: nil, ssr_timeout: nil,
                   profile_server_rendering_js_code: nil, raise_non_shell_server_rendering_errors: nil)
      self.renderer_url = renderer_url
      self.renderer_password = renderer_password
      self.server_renderer = server_renderer
      self.renderer_use_fallback_exec_js = renderer_use_fallback_exec_js
      self.prerender_caching = prerender_caching
      self.renderer_http_pool_size = renderer_http_pool_size
      self.renderer_http_pool_timeout = renderer_http_pool_timeout
      self.renderer_http_pool_warn_timeout = renderer_http_pool_warn_timeout
      self.tracing = tracing
      self.rendering_returns_promises = server_renderer == "NodeRenderer" ? rendering_returns_promises : false
      self.dependency_globs = dependency_globs
      self.excluded_dependency_globs = excluded_dependency_globs
      self.remote_bundle_cache_adapter = remote_bundle_cache_adapter
      self.ssr_pre_hook_js = ssr_pre_hook_js
      self.assets_to_copy = assets_to_copy
      self.renderer_request_retry_limit = renderer_request_retry_limit
      self.throw_js_errors = throw_js_errors
      self.ssr_timeout = ssr_timeout
      self.profile_server_rendering_js_code = profile_server_rendering_js_code
      self.raise_non_shell_server_rendering_errors = raise_non_shell_server_rendering_errors
    end
    # rubocop:enable Metrics/AbcSize

    def setup_config_values
      configure_default_url_if_not_provided
      validate_url
      validate_remote_bundle_cache_adapter
      setup_renderer_password
      setup_assets_to_copy
      setup_execjs_profiler_if_needed
    end

    def setup_execjs_profiler_if_needed
      return unless profile_server_rendering_js_code && server_renderer == "ExecJS"

      if ExecJS.runtime == ExecJS::Runtimes::Node
        ExecJS.runtime = ExecJS::ExternalRuntime.new(
          name: "Node.js (V8)",
          command: ["node --prof"],
          runner_path: "#{ExecJS.root}/support/node_runner.js",
          encoding: "UTF-8"
        )
      elsif ExecJS.runtime == ExecJS::Runtimes::V8
        ExecJS.runtime = ExecJS::ExternalRuntime.new(
          name: "V8",
          command: ["d8 --prof"],
          runner_path: "#{ExecJS.root}/support/v8_runner.js",
          encoding: "UTF-8"
        )
      else
        current_runtime = ExecJS.runtime.name
        message = <<~MSG
          You have set `profile_server_rendering_js_code` to true, but the current execjs runtime is #{current_runtime}.
          ExecJS profiler only supports Node.js (V8) or V8 runtimes.
          You can set the runtime by setting the `EXECJS_RUNTIME` environment variable to either `Node` or `V8`.
        MSG
        raise ReactOnRailsPro::Error, message
      end
    end

    def node_renderer?
      ReactOnRailsPro.configuration.server_renderer == "NodeRenderer"
    end

    private

    def setup_assets_to_copy
      self.assets_to_copy = (Array(assets_to_copy) if assets_to_copy.present?)
    end

    def configure_default_url_if_not_provided
      self.renderer_url = renderer_url.presence || DEFAULT_RENDERER_URL
    end

    def validate_url
      URI(renderer_url)
    rescue URI::InvalidURIError => e
      message = "Unparseable ReactOnRailsPro.config.renderer_url #{renderer_url} provided.\n#{e.message}"
      raise ReactOnRailsPro::Error, message
    end

    def validate_remote_bundle_cache_adapter
      if !remote_bundle_cache_adapter.nil? && !remote_bundle_cache_adapter.is_a?(Module)
        raise ReactOnRailsPro::Error, "config.remote_bundle_cache_adapter can only have a module or class assigned"
      end

      return unless remote_bundle_cache_adapter.is_a?(Module)

      unless remote_bundle_cache_adapter.methods.include?(:build)
        raise ReactOnRailsPro::Error,
              "config.remote_bundle_cache_adapter must have a class method named 'build'"
      end

      unless remote_bundle_cache_adapter.methods.include?(:fetch)
        raise ReactOnRailsPro::Error,
              "config.remote_bundle_cache_adapter must have a class method named 'fetch'" \
              "which takes a single named String parameter 'zipped_bundles_filename'" \
              "and returns the zipped file as a string if fetch attempt is successful & nil if not"
      end

      unless remote_bundle_cache_adapter.methods.include?(:upload) # rubocop:disable Style/GuardClause
        raise ReactOnRailsPro::Error,
              "config.remote_bundle_cache_adapter must have a class method named 'upload'" \
              "which takes a single named Pathname parameter 'zipped_bundles_filepath' & returns nil"
      end
    end

    def setup_renderer_password
      return if renderer_password.present?

      uri = URI(renderer_url)
      self.renderer_password = uri.password
    end
  end
end
